### PR TITLE
Add Proxy labels to support both GE and GG translation modes

### PR DESCRIPTION
**Pull request by jenshu**
Since we now need to have different translation logic between Gloo Edge and Gloo Gateway due to requirements in the Gateway API spec, we need a way to know which "translation mode" to use. Added labels to the Proxy indicating which translator (GE or GG) created the Proxy, so we know which translation logic to use.

### CODE REVIEW

File: projects/gloo/pkg/translator/route_config.go [hunk 1]
```diff
@@ -627,8 +639,29 @@ func setEnvoyPathMatcher(ctx context.Context, in *matchers.Matcher, out *envoy_c
...
-			Prefix: path.Prefix,
+		if h.proxy.GetTranslationMode() == v1.Proxy_TRANSLATION_MODE_GATEWAY {
+			// Gateway API spec treats each path segment as a full word, i.e. a request with path /abcd should not match
+			// the matcher prefix /abc (but request to /abc/def should match) so we must use PathSeparatedPrefix unless
+			// the matcher path ends in / in which case we need to use Prefix (PathSeparatedPrefix cannot end in /)
+			// See https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.PathMatchType - PathPrefix
+			// and https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.HTTPRouteRule - matches field
+			// for how matching should be done
+
+			// TODO: we may need more validation here.
```
Comment (Nadine2016) [inline projects/gloo/pkg/translator/route_config.go:650]: do we want to leave the To do in here?

Comment (jenshu) [reply projects/gloo/pkg/translator/route_config.go:650]: we can leave it there for now, i don't think it's resolved yet

File: projects/clusteringress/pkg/translator/translator_syncer.go [hunk 2]
```diff
@@ -36,7 +37,7 @@ type translatorSyncer struct {
 var (
 	// labels used to uniquely identify Proxies that are managed by the Gloo controllers
 	proxyLabelsToWrite = map[string]string{
-		"created_by": "gloo-knative",
+		glooutils.TranslatorKey: "gloo-knative",
```
Comment (lgadban) [inline projects/clusteringress/pkg/translator/translator_syncer.go:40]: do we care about adding `gloo-knative` to `utils/translators.go`?

File: projects/ingress/pkg/translator/translator_syncer.go [hunk 3]
```diff
@@ -34,7 +35,7 @@ type translatorSyncer struct {
 var (
 	// labels used to uniquely identify Proxies that are managed by the Gloo controllers
 	proxyLabelsToWrite = map[string]string{
-		"created_by": "gloo-ingress",
+		glooutils.TranslatorKey: "gloo-ingress",
```
Comment (lgadban) [inline projects/ingress/pkg/translator/translator_syncer.go:38]: same here; any value in moving this value to util?

**Review by lgadban — COMMENTED** [review]
general question, how do we plan on merging?
so we will merge this first, then the v2.0.x branch will rebase/merge these changes into it?

File: projects/gloo/pkg/translator/route_config.go [hunk 4]
```diff
@@ -627,8 +640,29 @@ func setEnvoyPathMatcher(ctx context.Context, in *matchers.Matcher, out *envoy_c
...
+			// envoy uses this regex: "^[^?#]+[^?#/]$"
+			if strings.HasSuffix(path.Prefix, "/") {
+				out.PathSpecifier = &envoy_config_route_v3.RouteMatch_Prefix{
+					Prefix: path.Prefix,
+				}
+			} else {
+				out.PathSpecifier = &envoy_config_route_v3.RouteMatch_PathSeparatedPrefix{
+					PathSeparatedPrefix: path.Prefix,
+				}
+			}
```
Comment (davidjumani) [inline projects/gloo/pkg/translator/route_config.go:661]: For posterity : https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routematch-path-separated-prefix
```
Expect the value to not contain ? or # and not to end in /
```

**Comment by jenshu** [comment]
> general question, how do we plan on merging? so we will merge this first, then the v2.0.x branch will rebase/merge these changes into it?

initially i was developing against the v2.0.x branch and was planning to merge this current PR into [this PR](https://github.com/solo-io/gloo/pull/9076), which would then merge into `v2.0.x`, and then open a new PR to merge `v2.0.x` into `main`. But since we are not planning to do further development against the 2.0 branch, the new plan is once this current PR is reviewed, just to close this PR and the [other 2.0 PR](https://github.com/solo-io/gloo/pull/9076), and instead just directly review/merge [the new PR](https://github.com/solo-io/gloo/pull/9117) into main.

basically this PR was only opened for ease of review. the branch that https://github.com/solo-io/gloo/pull/9117 is based on contains the same content as the branch the current PR (https://github.com/solo-io/gloo/pull/9105) is based on, except for changelog entries.

File: projects/clusteringress/pkg/translator/translator_syncer.go (hunk 2 above)
Comment (jenshu) [reply projects/clusteringress/pkg/translator/translator_syncer.go:40]: we decided to make the minimal required changes for now and not refactor all this; also i wasn't sure why knative seems to use 2 different labels: `gloo-knative` here and  `gloo-knative-translator` [here](https://github.com/solo-io/gloo/blob/d2c904f1fc066f2251e7aa91fbe34e0dcfcccf9f/projects/knative/pkg/translator/translator_syncer.go#L119)

File: projects/ingress/pkg/translator/translator_syncer.go (hunk 3 above)
Comment (jenshu) [reply projects/ingress/pkg/translator/translator_syncer.go:38]: same as above

**Comment by jenshu** [comment]
closing this out since it has been reviewed and verbally approved, and is included in https://github.com/solo-io/gloo/pull/9117
